Technical/Logging/The-Tazama-Logging-Framework.md

```diff
@@ -1 +1,405 @@
+
+
+###### Event-Sidecar
+A sample [.env](https://github.com/frmscoe/event-sidecar/blob/feb5e53b0801f60fa746e1720349cfb8c09e6c2b/.env.example) is provided.
+
+| Variable Name  | Purpose                              | Example                    |
+|----------------|--------------------------------------|----------------------------|
+| PORT           | Specifies the port number to use     | `8080`                     |
+| NATS_SERVER    | Specifies the NATS server address    | `nats://localhost:4222`    |
+| NATS_SUBJECT   | Defines the NATS subject/topic name  | `Lumberjack`               |
+
+> [!CAUTION]
+> Ensure the sidecar and lumberjack are pointing to the same instance of [NATS] and have the same subject
+
+###### Processor
+In your processor, add the [frms-coe-lib] as a dependency:
+
+```sh
+npm i @frmscoe/frms-coe-lib
+```
+
+In your processor, create an instance of a `LoggerService` and specify the [event-sidecar] address:
+
+```ts
+import { LoggerService } from '@frmscoe/frms-coe-lib';
+
+const logger = new LoggerService("localhost:8080");
+```
+> [!CAUTION]
+> Ensure the sidecar AND [lumberjack] start up before the processor so that logs are not missed.
+
+##### Without Event-Sidecar and Lumberjack
+In your processor, add the [frms-coe-lib] as a dependency:
+
+```sh
+npm i @frmscoe/frms-coe-lib
+```
+
+Configure your environment:
+| Variable Name         | Purpose                               | Example             |
+|-----------------------|---------------------------------------|---------------------|
+| ELASTIC_SEARCH_VERSION| Specifies the Elasticsearch version   | `8.11`              |
+| ELASTIC_HOST          | Specifies the Elasticsearch host       | `http://localhost:9200` |
+| ELASTIC_USERNAME      | Username for Elasticsearch authentication | `elastic_user`   |
+| ELASTIC_PASSWORD      | Password for Elasticsearch authentication | `secretpassword` |
+
+In your processor, create an instance of a `LoggerService`:
+
+```ts
+import { LoggerService } from '@frmscoe/frms-coe-lib';
+
+const logger = new LoggerService();
+```
+#### Usage
+
+You may specify a `LOGSTASH_LEVEL` environment variable in your processor to control your logs. The logging system supports multiple levels of severity, listed in ascending order of importance: `trace`, `debug`, `info`, `warn`, `error`, and `fatal`. When configuring the logging level, every log event with that level or a higher severity will be logged. For example, if you set the logging level to `info`, all log events at `info`, `warn`, `error`, and `fatal` levels will be logged. This ensures that important information and higher severity events are captured while filtering out less critical details.
+
+After initialising your logger, you may use the methods available to the `LoggerService` instance.
+
+##### Logger Service 
+
+
+###### Function Signature:
+
+The custom logger function has the following signature for `trace`, `debug`, `log` and `warn` log levels:
```

Comment:
`warn` is missing below.

Also, out of curiosity, we are there separate functions, instead of a log-level "switch" as a parameter to a single function?